src/transformers/pipelines/token_classification.py

```diff
@@ -245,38 +283,84 @@ def _forward(self, model_inputs):
+        # Check whether offset_mapping comes from tokenizer or manual input
+        if "offset_mapping" in all_outputs[0].keys():
+            x = all_outputs[0]["offset_mapping"]
+            if self.framework == "tf":
+                if tf.is_tensor(x):
+                    keys.append("offset_mapping")
+            else:
+                if torch.is_tensor(x):
+                    keys.append("offset_mapping")
+        outputs = {}
+        for k in keys:
+            outputs[k] = np.array([model_outputs[k][0].numpy() for model_outputs in all_outputs])
+        logits = outputs.pop("logits")
         maxes = np.max(logits, axis=-1, keepdims=True)
         shifted_exp = np.exp(logits - maxes)
         scores = shifted_exp / shifted_exp.sum(axis=-1, keepdims=True)
 
-        if self.framework == "tf":
-            input_ids = input_ids.numpy()
-            offset_mapping = offset_mapping.numpy() if offset_mapping is not None else None
-
-        pre_entities = self.gather_pre_entities(
-            sentence, input_ids, scores, offset_mapping, special_tokens_mask, aggregation_strategy
-        )
-        grouped_entities = self.aggregate(pre_entities, aggregation_strategy)
-        # Filter anything that is in self.ignore_labels
-        entities = [
-            entity
-            for entity in grouped_entities
-            if entity.get("entity", None) not in ignore_labels
-            and entity.get("entity_group", None) not in ignore_labels
-        ]
-        return entities
+        input_ids = outputs.pop("input_ids")
+        special_tokens_mask = outputs.pop("special_tokens_mask")
+        offset_mapping = outputs.pop("offset_mapping", None)
+
+        if offset_mapping is None:
+            offset_mapping = all_outputs[0].pop("offset_mapping", None)
+            offset_mapping = offset_mapping if offset_mapping is not None else [None]
+
+        aggregated_entities = []
+        for i in range(num_chunks):
+            pre_entities = self.gather_pre_entities(
+                sentence,
+                input_ids[i],
+                scores[i],
+                offset_mapping[i],
+                special_tokens_mask[i],
+                aggregation_strategy,
+            )
+            grouped_entities = self.aggregate(pre_entities, aggregation_strategy)
+            # Filter anything that is in self.ignore_labels
+            entities = [
+                entity
+                for entity in grouped_entities
+                if entity.get("entity", None) not in ignore_labels
+                and entity.get("entity_group", None) not in ignore_labels
+            ]
+            aggregated_entities.extend(entities)
+
+        if num_chunks > 1:
+            aggregated_entities = self.aggregate_entities(aggregated_entities)
+        return aggregated_entities
+
+    def aggregate_entities(self, entities):
+        if len(entities) == 0:
+            return entities
+        entities = sorted(entities, key=lambda x: x["start"])
```

Comment:
@Narsil, here is an example if you remove `sorted()` and you have entities in the overlapping part.
I took the Wikipedia article of Elon Musk (we have lot of entities to visualize) and comment this line:
`# entities = sorted(entities, key=lambda x: x["start"])`
The aggregation function is not able to work anymore now. We have duplicates entities. Like SolarCity, it is present twice:
![image](https://user-images.githubusercontent.com/74506016/224751279-6bf8d374-2f1e-407b-b9da-1fb8539efeb7.png)

Reply:
@Narsil, for performance reasons regarding `aggregate_overlapping_entities()`, I don't think we can do better in pure python. I understand that at first glance this function seems to be not optimized due to the `sorted()` method but in fact it is.

The number of entities in overlapping parts can be different between chunks. We can't be sure we don't have an overlapping entity until we look at the next one. Entities must be sorted for this.

For some specific applications where you need the full capability of the transformers with the widest context, you can choose a stride number greater than `model_max_length / 2` and in that case you will have the same entity in many chunks and it is important to tackle this case.

Even if you don't want to sort the entities, you don't know where the overlapping entities are and that will end up with different loops (at least O(n²) complexity) and method calls, which is time consuming. Whereas now with `sorted()` + the for loop we have an average time complexity of O(2n), because `sorted()` is already deeply optimized.

As you can see below, even for 100000 entities retrieved in one text, the function only takes 40ms to aggregate them. The calculation time of all these chunks to extract these 100000 entities will be much more important than these 40ms.
![image](https://user-images.githubusercontent.com/74506016/224982687-8479da56-14da-499a-91ec-50dae4fd0083.png)